### ISSUE TITLE
Could not load file or assembly 'System.Collections.Immutable with 3.0.3

### ISSUE THREAD

**Issue by richardjharding** [SteveDunn/Vogen#275]
### Describe the bug

After upgrading to 3.03  I started getting build errors:

```
CSC : warning CS8032: An instance of analyzer Vogen.Rules.AddValidationAnalyzer cannot be created from C:\Users\richa\.nuget\packages\vogen\3.0.3\analyzers\dotnet\cs\Vogen.dll : Could not load file or assembly 'System.Collections.Immutable, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Vogen.Rules.DoNotUseDefaultAnalyzer cannot be created from C:\Users\richa\.nuget\packages\vogen\3.0.3\analyzers\dotnet\cs\Vogen.dll : Could not load file or assembly 'System.Collections.Immutable, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Vogen.Rules.DoNotUseNewAnalyzer cannot be created from C:\Users\richa\.nuget\packages\vogen\3.0.3\analyzers\dotnet\cs\Vogen.dll : Could not load file or assembly 'System.Collections.Immutable, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8032: An instance of analyzer Vogen.Rules.DoNotUseReflectionAnalyzer cannot be created from C:\Users\richa\.nuget\packages\vogen\3.0.3\analyzers\dotnet\cs\Vogen.dll : Could not load file or assembly 'System.Collections.Immutable, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified..
1>CSC : warning CS8784: Generator 'ValueObjectGenerator' failed to initialize. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'FileNotFoundException' with message 'Could not load file or assembly 'System.Collections.Immutable, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.'

```

Downgrading to 3.0.1 solves the issue, explicitly adding the system.collections.immutable package did not fix the error

Did I miss a step in the upgrade?

### Steps to reproduce

updated the nuget package to 3.0.3 and clean and rebuild

### Expected behaviour

successful compilation

**Comment by SteveDunn** [comment]
Thanks for the bug report @richardjharding - what framework are you targetting?

**Comment by SteveDunn** [comment]
Also, what version of Visual Studio are you using? There was a bug in Visual Studio 2022 Preview 2.1 (and lower) which meant that analyzers would try to load the wrong version of System.Collections.Immutable

**Comment by SteveDunn** [comment]
I think the version of Visual Studio is the key here. You need that latest version of either 2022 or 2022 Preview

**Comment by richardjharding** [comment]
I'm on Microsoft Visual Studio Professional 2022 (64-bit) - Preview Version 17.4.0 Preview 2.0

any other logs that might be helpful?

**Comment by richardjharding** [comment]
seems I'm a couple versions behind on preview - I'll update and see if that helps - preview 6.0 is out

**Comment by SteveDunn** [comment]
> I'm on Microsoft Visual Studio Professional 2022 (64-bit) - Preview Version 17.4.0 Preview 2.0
> 
> any other logs that might be helpful?

Aha! I had the exact same thing and I literally spent **3 solid days** trying to suss out what the issue was. It is actually a bug in Visual Studio that was fixed in 17.4 Preview 3!

Vogen has a dependency on Microsoft.CodeAnalysis.CSharp, version 4.3.1, but this has a dependency on System.Collections.Immutable version 6.0. Thing is, it shouldn't - it should reference 5.0 because that's maximum version for netstandard. All analyzers should only target netstandard. It's all bit strange; if you use reflector to load that assembly, you'll see, even for this 'fixed' version, that it still has the wrong dependency:

![image](https://user-images.githubusercontent.com/263416/200136637-71a38e08-785d-44b7-8ad9-45e2e9f686f9.png)

Anyway, MS fixed it in Visual Studio itself. The Roslyn team helped me immensely on Discord!

I've never been more glad to see something fixed (albeit magically as far as my knowledge extends!)

**Comment by richardjharding** [comment]
yep that looks to have been it thanks! - upgrading to the latest preview solved it
